reana_server/fetcher.py

```diff
@@ -364,38 +372,52 @@ def _get_github_fetcher(parsed_url: ParsedUrl, output_dir: str) -> WorkflowFetch
         raise ValueError("Malformed GitHub URL")
     if tree_or_blob and not git_ref:
         raise ValueError("Branch or commit ID not provided in GitHub URL")
-    if tree_or_blob == "blob" and not path:
-        raise ValueError("Specification file not provided in GitHub URL")
     if tree_or_blob == "blob":
-        _, extension = os.path.splitext(path)
-        if extension not in WORKFLOW_SPEC_EXTENSIONS:
-            raise ValueError("GitHub URL points to an invalid specification file")
+        raise ValueError(
+            "GitHub URL points directly to a file. "
+            "Use the 'spec' argument to specify a particular specification file"
+        )
     if tree_or_blob == "tree" and path:
         raise ValueError("GitHub URL points to a directory")
 
     repository_url = ParsedUrl(f"https://github.com/{username}/{repository}.git")
-    return WorkflowFetcherGit(repository_url, output_dir, git_ref, spec=path)
+    return WorkflowFetcherGit(repository_url, output_dir, git_ref, spec)
 
 
-def get_fetcher(launcher_url: str, output_dir: str) -> WorkflowFetcherBase:
+def get_fetcher(
+    launcher_url: str, output_dir: str, spec: Optional[str] = None
+) -> WorkflowFetcherBase:
     """Select the correct workflow fetcher based on the given URL.
 
     :param launcher_url: URL of the workflow specification.
     :param output_dir: Directory where all the data fetched will be saved.
+    :param spec: Optional path to the workflow specification.
     :returns: Workflow fetcher.
     """
     parsed_url = ParsedUrl(launcher_url)
 
     if parsed_url.scheme not in FETCHER_ALLOWED_SCHEMES:
         raise ValueError("URL scheme not allowed")
 
+    if spec:
+        _, spec_ext = os.path.splitext(spec)
+        if spec_ext not in WORKFLOW_SPEC_EXTENSIONS:
+            raise ValueError(
+                "The provided specification doesn't have a valid file extension"
+            )
+
     if parsed_url.extension == ".git":
-        return WorkflowFetcherGit(parsed_url, output_dir)
+        return WorkflowFetcherGit(parsed_url, output_dir, spec=spec)
```

Comment:
Extremely minor: is there a reason not to pass `spec` as a positional argument here like for the other classes?

Reply:
The third parameter of the initializer of `WorkflowFetcherGit` is an optional reference to a Git branch/commit (`git_ref`). Should I swap `git_ref` and `spec` so that the initialization of `WorkflowFetcherGit` is consistent with the initialization of the other types of fetchers?

Reply:
I see, I've missed that there is additional `git_ref` parameter in between `spec` and others for `WorkflowFetcherGit`.. I think there is no need to change anything here, thanks for explaining